商品購入機能

hi-roto opened the pull request:
# What
　商品購入機能

# Why
　商品購入機能を実装するため

# gif一覧
[きちんと商品購入ができる](https://gyazo.com/2630be7bec8a741eec1b12eb960069a6)
---
[決まったユーザー以外がURLに直接打ち込んでも商品購入ページに遷移しない](https://gyazo.com/bb4b6fce4dade590ad4ecdf9fbd54314)
---
[バリデーションがきちんと働いている](https://gyazo.com/5746eebf4a4dd3896941f610bdff3f07)
---
[テストコードの実行](https://gyazo.com/4dbe483ee2cc9c03235666eeb18b9f45)
---

hi-roto committed:
purchase_recordモデルとshipping_addressモデルの作成とマイグレーションファイルの作成

hi-roto committed:
purchase_recordのビューファイルの設置

hi-roto committed:
ルーティングの記載

hi-roto committed:
formオブジェクトの作成とコントローラの記載

hi-roto committed:
payjsの実装

hi-roto committed:
ビューファイルの修正

hi-roto committed:
テストコードの実装

chisesakai commented:
レビュー依頼ありがとうございます！
キャプチャ拝見させていただきました。
要件がまとまっていて素晴らしいです！
#### 追加で挙動のキャプチャをのご提出をお願い致します
実際の画面や機能ができているか確認させていただきたいので、下記キャプチャ画像をご提出いただけますか？
- 出品者であるログイン状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画(編集機能要件より)
- ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（詳細機能要件より)

その後、再度レビュー依頼フォームにてご提出をお願いいたします。

hi-roto committed:
コントローラとビューに「売れた商品の編集や削除できない」機能を追記

hi-roto commented:
ご指摘の機能が未実装であったため、修正・追加させていただきました。
再度のレビューよろしくお願いします。

[特定の条件下でないと編集ページに遷移できない](https://gyazo.com/f33e367587afe39c0654f83c73c539c0)

shodai-ito commented on config/routes.rb:
```diff
@@ -1,5 +1,11 @@
   devise_for :users
   root 'items#index'
-  resources :items, only: [:new, :create, :show, :edit, :update, :destroy]
+  resources :items, only: [:new, :create, :edit, :update, :destroy]
```
6行目に記述がありますのでこちらは削除しましょう！

hi-roto replied:
[修正しました](https://github.com/hi-roto/furima-32883/pull/8/commits/b3c81a86ed5f6c207663a016781daade931d0736)

shodai-ito commented on app/models/purchase_record_shipping_address.rb:
```diff
@@ -0,0 +1,19 @@
+
+  VALID_POSTAL_CODE = /\A\d{3}-\d{4}\z/
+  with_options presence: true do
+    validates :token
+    validates :postal_code, format: { with: VALID_POSTAL_CODE, message: 'Input correctly' }
+    validates :prefecture_id, numericality: { other_than: 1, message: 'Select' }
+    validates :city, :house_number
+    validates :phone_number, numericality: { only_integer: true, message: 'Input only number' }, length: { in: 10..11 }
```
こちらuser_idとitem_idに対してもバリデーションを設定しましょう！
理由は、  今まではモデル間のアソシエーションを組むことで、自動的にpresence: trueが働いておりました。  ただ、フォームオブジェクトでは直接アソシエーションを組んでいないためにバリデーションが必要となります。

hi-roto replied:
[修正しました](https://github.com/hi-roto/furima-32883/pull/8/commits/b872c9d8a3e9b4a40594c3483ddc0b0b6946a519)

shodai-ito commented on spec/models/purchase_record_shipping_address_spec.rb:
```diff
@@ -0,0 +1,58 @@
+        @purchase_record_shipping_address.postal_code = ''
+        @purchase_record_shipping_address.valid?
+        expect(@purchase_record_shipping_address.errors.full_messages).to include("Postal code can't be blank")
+      end
```
郵便番号にはハイフンが必要であることをテストしましょう！

shodai-ito commented on spec/models/purchase_record_shipping_address_spec.rb:
```diff
@@ -0,0 +1,58 @@
+        expect(@purchase_record_shipping_address.errors.full_messages).to include("Postal code can't be blank")
+      end
+      it '都道府県が空では保存できない' do
+        @purchase_record_shipping_address.prefecture_id = '1'
```
prefecture_idはinteger型ですので''を削除しましょう！

shodai-ito commented on spec/models/purchase_record_shipping_address_spec.rb:
```diff
@@ -0,0 +1,58 @@
+        @purchase_record_shipping_address.phone_number = '123456789101'
+        @purchase_record_shipping_address.valid?
+        expect(@purchase_record_shipping_address.errors.full_messages).to include('Phone number is too long (maximum is 11 characters)')
+      end
```
こちらuser_idとitem_idに対してもテストを記述しましょう！

shodai-ito commented on spec/models/purchase_record_shipping_address_spec.rb:
```diff
@@ -0,0 +1,58 @@
+
+RSpec.describe PurchaseRecordShippingAddress, type: :model do
+  before do
+    @purchase_record_shipping_address = FactoryBot.build(:purchase_record_shipping_address)
```
商品購入には商品と購入者の情報が必要です。
before doの中に記述しましょう！

hi-roto replied:
こちらは、
```ruby
FactoryBot.define do
  factory :purchase_record_shipping_address do
    #（省略）
    user_id { Faker::Number.digit}
    item_id { Faker::Number.digit}
  end
end
```
と記載して、
 `@purchase_record_shipping_address = FactoryBot.build(:purchase_record_shipping_address)`
で修正しました。

ky3090 replied:
こちらですが、以下を参考に実装されてください。
理由としては、よりアプリケーションの実装に近い形でテストを行うためです。

```
 1. ユーザーのダミーデータをcreateメソッドで作成し、変数格納
 2. アイテムのダミーデータをcreateメソッドで作成し、変数格納
 3. @purchase_record_shipping_address = FactoryBot.build(:purchase_record_shipping_address)にitem_idとuser_idをキーとし、そのバリューとして1,2で定義した変数からidを抜き出す記述を定義する
 4. テストコード実行の際、遅延によってエラーが起きるので、エラー回避のためsleepメソッドを追記する。
```

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

再度レビュー依頼フォームのご提出をよろしくお願いいたします。

hi-roto committed:
ルーティングの修正

hi-roto committed:
バリデーションの追記

hi-roto committed:
テストコードの修正

１　郵便番号にハイフンが必要であること
２　都道府県のテストコードの値に''を削除
３　購入者が不明だと保存できないこと
４　商品が不明だと購入できない
を修正し、追加

hi-roto commented:
レビューありがとうございます。
該当箇所の修正を完了しました。
[テストコードの実行結果です](https://gyazo.com/e75f3a05d1e60badabbf9c9792ec60a8)

ky3090 commented on spec/factories/purchase_record_shipping_addresses.rb:
```diff
@@ -0,0 +1,12 @@
+    phone_number { Faker::Number.number(digits: 10) }
+    token { 'tok_abcdefghijk00000000000000000' }
+    user_id { Faker::Number.digit}
+    item_id { Faker::Number.digit}
```
```
 user_id { Faker::Number.digit}
    item_id { Faker::Number.digit}
```

先ほどのコメントの修正を行うことで、上記の記述は不要となります。

ky3090 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

hi-roto committed:
テストコードの修正

hi-roto commented:
ご指摘の点を修正しました。
レビューよろしくお願いします。
[テストコード実行結果](https://gyazo.com/c8caecb762e24095743d6456ae5f213a)

chisesakai commented on app/views/items/show.html.erb:
```diff
@@ -24,14 +24,12 @@
     <%= link_to "商品の編集", edit_item_path, method: :get, class: "item-red-btn" %>
     <p class="or-text">or</p>
     <%= link_to "削除", item_path(@item.id), method: :delete, class:"item-destroy" %>
-  <% elsif user_signed_in? %>
-    <%# 商品が売れていない場合はこちらを表示しましょう %>
-    <%= link_to "購入画面に進む", "#" ,class:"item-red-btn"%>
-    <%# //商品が売れていない場合はこちらを表示しましょう %>
+  <% elsif user_signed_in? && @item.purchase_record.blank? %>
+    <%= link_to "購入画面に進む", purchase_records_path(@item.id) ,class:"item-red-btn"%>
   <%end%>
```
```Ruby
  <% if (user_signed_in? && current_user.id == @item.user.id) && @item.purchase_record.blank? %>
    <%= link_to "商品の編集", edit_item_path, method: :get, class: "item-red-btn" %>
    <p class="or-text">or</p>
    <%= link_to "削除", item_path(@item.id), method: :delete, class:"item-destroy" %>
  <% elsif user_signed_in? && @item.purchase_record.blank? %>
    <%= link_to "購入画面に進む", purchase_records_path(@item.id) ,class:"item-red-btn"%>
  <%end%>
```
以下のような構造でリファクタリングしてみましょう。
```ruby
if "ログインしていれば" かつ "商品は売れていなければ" 
  if "現在ログインしているユーザーが商品の出品者であれば"
    商品の編集
    削除
  else
    購入画面に進む
  end
end
```

ポイント
１．大きな条件として「"ログインしていれば" かつ "商品は売れていなければ" 」
　こちらの条件で「商品の編集」「削除」「購入画面に進む」が表示されるかを制御しましょう。

２．１の条件の中でさらに条件を追加します。
　「"現在ログインしているユーザーが商品の出品者であれば"」
　こちらの条件で「商品の編集」「削除」を表示させるか、
  「購入画面に進む」を表示させるかを制御しましょう。

chisesakai commented on spec/models/purchase_record_shipping_address_spec.rb:
```diff
@@ -0,0 +1,76 @@
+  end
+
+  describe '商品購入機能' do
+    context '商品購入ができる時' do
```
```
    context '商品購入ができる時' do
```
正常系パターンのテスト項目を追記しましょう！
```
・建物名が抜けていても登録できること
```

chisesakai commented on spec/models/purchase_record_shipping_address_spec.rb:
```diff
@@ -0,0 +1,76 @@
+      it '電話番号が空では保存できない' do
+        @purchase_record_shipping_address.phone_number = nil
+        @purchase_record_shipping_address.valid?
+        expect(@purchase_record_shipping_address.errors.full_messages).to include("Phone number can't be blank")
```
電話番号に関しまして、下記のテストコードの記述もお願いいたします。
```
・全角数字だと登録できないこと
```

chisesakai reviewed (commented):
学習お疲れさまです！レビュー依頼ありがとうございます。
修正点確認させていただきました。
#### 追加で3件コメントいたしましたので、ご確認お願いいたします。
あともう一息かと思いますので、引き続き一緒に頑張っていきましょう！
修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！

hi-roto committed:
条件分岐のリファクタリング

hi-roto committed:
テストコードの修正

hi-roto commented:
指摘頂いた箇所を訂正させていただきました。
再度、レビューよろしくお願いします。
[テストコード実行結果](https://gyazo.com/4a4614185fd9beb184564604da001d15)

Katsuno0814 reviewed (commented):
修正とレビュー依頼ありがとうございます。

問題なく実装できていることが確認できましたので、
LGTMとさせていただきます！

こちらのブランチは忘れずにMergeを行っていただき、
ローカルのMasterにPullしておきましょう。

ここまでの実装、本当にお疲れ様でした。
引き続き、応援しております。
